File: server/plugins.py
"""Plugin discovery + lifecycle.

Two plugin sources, with different trust levels:

* **Built-in** — `<repo>/plugins/<name>/`, shipped with Grimoire. Trusted code;
  the on-topic ones (see `DEFAULT_ENABLED_BUILTINS`) are enabled by default, the
  rest ship one toggle away. Each can be enabled/disabled from Settings.
* **Vault** — `<vault>/plugins/<name>/`, user-installed and synced with the
  vault like any other content. Because a vault plugin is arbitrary JavaScript
  executed in the app's origin, vault plugins are **disabled by default** and
  must be enabled explicitly, one by one, from Settings. The UI shows a
  warning when enabling one.

A plugin is a directory with a `plugin.json` manifest:

    {
      "name": "kanban",                  // must match the directory name
      "version": "1.0.0",
      "description": "Kanban boards from ```kanban fences",
      "client": "client.js",             // entry, loaded as an ES module
      "styles": "style.css"              // optional stylesheet
    }

Enablement state lives in `.grimoire/plugins.json` (vault-local, not synced
content — each device/host decides what runs). Asset serving is path-confined
to the plugin's own directory.
"""
from __future__ import annotations

import json
import re
from pathlib import Path

from . import config

BUILTIN_DIR = config.ROOT / "plugins"
_NAME_RE = re.compile(r"^[a-z][a-z0-9-]{0,40}$")

# Which built-ins are ON out of the box. The bar: does it make the *notes /
# knowledge* richer? Content renderers (math, diagrams, boards) are invisible
# until you use their syntax, and vault-stats reports on the vault itself — all
# on-topic. The productivity widgets (pomodoro timer, writing-streak heatmap,
# daily word goal) are genuinely useful but off-topic sidebar furniture; they
# ship enabled-able, not enabled, so a fresh vault looks like a focused tool
# rather than a kitchen sink. Everything here is still one toggle away.
DEFAULT_ENABLED_BUILTINS = {"katex", "mermaid", "kanban", "vault-stats"}


def _state_path() -> Path:
    return config.grimoire_dir() / "plugins.json"


def _load_state() -> dict:
    try:
        return json.loads(_state_path().read_text(encoding="utf-8"))
    except Exception:
        return {}


def _save_state(state: dict) -> None:
    _state_path().parent.mkdir(parents=True, exist_ok=True)
    _state_path().write_text(json.dumps(state, indent=2), encoding="utf-8")


def vault_dir() -> Path:
    return config.VAULT / "plugins"


def _read_manifest(pdir: Path, source: str) -> dict | None:
    """Load + validate one plugin's manifest. Returns None for anything broken —
    a malformed plugin must never take the app down."""
    mf = pdir / "plugin.json"
    if not mf.is_file():
        return None
    try:
        data = json.loads(mf.read_text(encoding="utf-8"))
    except Exception:
        return None
    name = data.get("name", "")
    if name != pdir.name or not _NAME_RE.match(name):
        return None                      # manifest must match its directory
    client = data.get("client", "client.js")
    if not (pdir / client).is_file():
        return None
    return {
        "name": name,
        "version": str(data.get("version", "0.0.0")),
        "description": str(data.get("description", "")),
        "source": source,
        "client": client,
        "styles": data.get("styles") if isinstance(data.get("styles"), str) else None,
    }


def discover() -> list[dict]:
    """All plugins from both sources with their effective enabled state.
    Built-ins default to enabled; vault plugins default to DISABLED (untrusted)."""
    state = _load_state()
    out: list[dict] = []
    seen: set[str] = set()
    for source, root in (("builtin", BUILTIN_DIR), ("vault", vault_dir())):
        if not root.is_dir():
            continue
        for pdir in sorted(root.iterdir()):
            if not pdir.is_dir() or pdir.name in seen:
                continue                 # builtin name wins over a vault clone
            m = _read_manifest(pdir, source)
            if not m:
                continue
            default_on = source == "builtin" and m["name"] in DEFAULT_ENABLED_BUILTINS
            m["enabled"] = bool(state.get(m["name"], {}).get("enabled", default_on))
            out.append(m)
            seen.add(pdir.name)
    return out


def set_enabled(name: str, enabled: bool) -> dict | None:
    """Persist enablement. Returns the plugin entry, or None if unknown."""
    plugin = next((p for p in discover() if p["name"] == name), None)
    if plugin is None:
        return None
    state = _load_state()
    state.setdefault(name, {})["enabled"] = enabled
    _save_state(state)
    plugin["enabled"] = enabled
    return plugin


def asset_path(name: str, rel: str) -> Path | None:
    """Resolve a plugin asset path, confined to that plugin's directory.
    Only enabled plugins serve assets (a disabled plugin's code never loads)."""
    plugin = next((p for p in discover() if p["name"] == name), None)
    if plugin is None or not plugin["enabled"]:
        return None
    root = (BUILTIN_DIR if plugin["source"] == "builtin" else vault_dir()) / name
    try:
        p = (root / rel).resolve()
        p.relative_to(root.resolve())    # raises ValueError on traversal
    except ValueError:
        return None
    return p if p.is_file() else None

SCAFFOLD_CLIENT = """/**
 * {name} — a Grimoire plugin.
 *
 * This skeleton was generated by "Create a plugin". It registers one palette
 * command and one sidebar panel; delete what you don't need. Full API docs:
 * docs/PLUGINS.md in the Grimoire repo.
 *
 * Enable it under Settings → Plugins (vault plugins are off until you opt in).
 */
export function activate(grimoire) {{
  grimoire.registerCommand({{
    icon: "🔌",
    name: "{name}: hello",
    run: () => grimoire.toast("Hello from {name}!"),
  }});

  grimoire.registerPanel({{
    id: "{name}",
    title: "🔌 {name}",
    render(el) {{
      el.textContent = "Edit plugins/{name}/client.js in your vault to build me.";
    }},
  }});
}}
"""


def scaffold(name: str) -> dict:
    """Write a hello-world vault plugin skeleton. Stays DISABLED until the user
    enables it in Settings — scaffolding must not grant execution by itself."""
    if not _NAME_RE.match(name):
        raise ValueError("plugin name must be lowercase letters/digits/hyphens")
    if any(p["name"] == name for p in discover()):
        raise ValueError(f"a plugin named {name!r} already exists")
    pdir = vault_dir() / name
    pdir.mkdir(parents=True, exist_ok=False)
    (pdir / "plugin.json").write_text(json.dumps({
        "name": name, "version": "0.1.0",
        "description": "My Grimoire plugin (edit plugin.json and client.js)",
        "client": "client.js",
    }, indent=2), encoding="utf-8")
    (pdir / "client.js").write_text(SCAFFOLD_CLIENT.format(name=name), encoding="utf-8")
    return {"name": name, "path": f"plugins/{name}", "enabled": False}

File: go/internal/api/plugins.go
package api

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// Plugin API: list, enable/disable, and path-confined asset serving.
//
// The trust split is the point: built-ins ship with the server and the on-topic
// ones are enabled by default; VAULT plugins are untrusted code that arrived
// with someone's notes, so they are discovered but DISABLED until the human
// turns them on, and a disabled plugin serves nothing at all.

var pluginNameRE = regexp.MustCompile(`^[a-z][a-z0-9-]{0,40}$`)

// defaultEnabledBuiltins are on out of the box. The bar: does it make the notes
// richer? Content renderers are invisible until you use their syntax; the
// productivity widgets ship enable-able, not enabled, so a fresh vault looks
// like a focused tool rather than a kitchen sink.
var defaultEnabledBuiltins = map[string]bool{
	"katex": true, "mermaid": true, "kanban": true, "vault-stats": true,
}

type pluginInfo struct {
	Name        string  `json:"name"`
	Version     string  `json:"version"`
	Description string  `json:"description"`
	Source      string  `json:"source"`
	Client      string  `json:"client"`
	Styles      *string `json:"styles"`
	Enabled     bool    `json:"enabled"`
	ClientURL   string  `json:"client_url"`
	StylesURL   *string `json:"styles_url"`
}

func (s *Server) pluginStatePath() string {
	return filepath.Join(s.Vault.Root, ".grimoire", "plugins.json")
}

func (s *Server) loadPluginState() map[string]bool {
	out := map[string]bool{}
	raw, err := os.ReadFile(s.pluginStatePath())
	if err != nil {
		return out
	}
	_ = json.Unmarshal(raw, &out)
	return out
}

func (s *Server) savePluginState(state map[string]bool) error {
	if err := os.MkdirAll(filepath.Dir(s.pluginStatePath()), 0o755); err != nil {
		return err
	}
	raw, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.pluginStatePath(), raw, 0o644)
}

// readManifest loads and validates one plugin. Anything broken returns nil — a
// malformed plugin must never take the app down.
func readManifest(dir, source string) *pluginInfo {
	raw, err := os.ReadFile(filepath.Join(dir, "plugin.json"))
	if err != nil {
		return nil
	}
	var data struct {
		Name        string `json:"name"`
		Version     any    `json:"version"`
		Description string `json:"description"`
		Client      string `json:"client"`
		Styles      any    `json:"styles"`
	}
	if json.Unmarshal(raw, &data) != nil {
		return nil
	}
	base := filepath.Base(dir)
	// the manifest must match its directory, or a plugin could claim another's
	// name and shadow it
	if data.Name != base || !pluginNameRE.MatchString(data.Name) {
		return nil
	}
	client := data.Client
	if client == "" {
		client = "client.js"
	}
	if info, err := os.Stat(filepath.Join(dir, client)); err != nil || info.IsDir() {
		return nil
	}
	version := "0.0.0"
	if v, ok := data.Version.(string); ok && v != "" {
		version = v
	}
	var styles *string
	if v, ok := data.Styles.(string); ok && v != "" {
		styles = &v
	}
	return &pluginInfo{
		Name: data.Name, Version: version, Description: data.Description,
		Source: source, Client: client, Styles: styles,
	}
}

func (s *Server) discoverPlugins() []pluginInfo {
	state := s.loadPluginState()
	var out []pluginInfo
	seen := map[string]bool{}

	for _, src := range []struct{ source, root string }{
		{"builtin", s.PluginDir},
		{"vault", filepath.Join(s.Vault.Root, "plugins")},
	} {
		if src.root == "" {
			continue
		}
		entries, err := os.ReadDir(src.root)
		if err != nil {
			continue
		}
		names := make([]string, 0, len(entries))
		for _, e := range entries {
			if e.IsDir() {
				names = append(names, e.Name())
			}
		}
		sort.Strings(names)
		for _, name := range names {
			if seen[name] {
				continue
			}
			info := readManifest(filepath.Join(src.root, name), src.source)
			if info == nil {
				continue
			}
			seen[name] = true
			// built-ins default per the on-topic list; vault plugins default
			// to DISABLED because they are untrusted code
			info.Enabled = src.source == "builtin" && defaultEnabledBuiltins[name]
			if v, ok := state[name]; ok {
				info.Enabled = v
			}
			info.ClientURL = "/plugins/" + name + "/" + info.Client
			if info.Styles != nil {
				u := "/plugins/" + name + "/" + *info.Styles
				info.StylesURL = &u
			}
			out = append(out, *info)
		}
	}
	if out == nil {
		out = []pluginInfo{}
	}
	return out
}

func (s *Server) listPlugins(w http.ResponseWriter, _ *http.Request) {
	writeJSON(w, http.StatusOK, s.discoverPlugins())
}

func (s *Server) enablePlugin(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	var body struct {
		Enabled bool `json:"enabled"`
	}
	_ = json.NewDecoder(r.Body).Decode(&body)

	found := false
	for _, p := range s.discoverPlugins() {
		if p.Name == name {
			found = true
			break
		}
	}
	if !found {
		writeErr(w, http.StatusNotFound, "no such plugin")
		return
	}
	state := s.loadPluginState()
	state[name] = body.Enabled
	if err := s.savePluginState(state); err != nil {
		writeErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"name": name, "enabled": body.Enabled})
}

type scaffoldIn struct {
	Name string `json:"name"`
}

// scaffoldPlugin writes a hello-world vault plugin, disabled until enabled.
func (s *Server) scaffoldPlugin(w http.ResponseWriter, r *http.Request) {
	var in scaffoldIn
	if err := json.NewDecoder(r.Body).Decode(&in); err != nil {
		writeErr(w, http.StatusBadRequest, "invalid json")
		return
	}
	name := strings.ToLower(strings.TrimSpace(in.Name))
	if !pluginNameRE.MatchString(name) {
		writeErr(w, http.StatusBadRequest,
			"name must be lowercase letters, digits and hyphens")
		return
	}
	dir := filepath.Join(s.Vault.Root, "plugins", name)
	if _, err := os.Stat(dir); err == nil {
		writeErr(w, http.StatusBadRequest, "plugin already exists")
		return
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		writeErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	manifest := map[string]any{
		"name": name, "version": "0.1.0",
		"description": "a new grimoire plugin", "client": "client.js",
	}
	raw, _ := json.MarshalIndent(manifest, "", "  ")
	if err := os.WriteFile(filepath.Join(dir, "plugin.json"), raw, 0o644); err != nil {
		writeErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	client := "// " + name + " — a grimoire plugin.\n" +
		"export function activate(api) {\n" +
		"  api.registerCommand('" + name + ": hello', () => alert('hello from " + name + "'));\n" +
		"}\n"
	if err := os.WriteFile(filepath.Join(dir, "client.js"), []byte(client), 0o644); err != nil {
		writeErr(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{
		"name": name, "source": "vault", "enabled": false,
		"path": "plugins/" + name,
	})
}

// servePluginAsset serves a plugin file, confined to that plugin's directory.
// A disabled plugin serves NOTHING: leaving assets reachable would let
// untrusted vault code run via a direct URL even while switched off.
func (s *Server) servePluginAsset(w http.ResponseWriter, r *http.Request) {
	name := r.PathValue("name")
	rel := r.PathValue("rel")

	var plugin *pluginInfo
	all := s.discoverPlugins()
	for i := range all {
		if all[i].Name == name {
			plugin = &all[i]
			break
		}
	}
	if plugin == nil || !plugin.Enabled {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	root := s.PluginDir
	if plugin.Source == "vault" {
		root = filepath.Join(s.Vault.Root, "plugins")
	}
	base := filepath.Join(root, name)
	target := filepath.Clean(filepath.Join(base, rel))
	// path confinement: a plugin may only serve its own files
	if target != base && !strings.HasPrefix(target, base+string(filepath.Separator)) {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	info, err := os.Stat(target)
	if err != nil || info.IsDir() {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	http.ServeFile(w, r, target)
}
